### PR TITLE
Support parameters for base graphics

### PR DESCRIPTION
**Pull request by frosch123**
## Motivation / Problem

OpenGFX2 features a number of parameters to configure detail looks:
image

Currently this requires an extra NewGRF as work-around.

## Description

* Base graphics parameters are the NewGRF parameters of the Extra-GRF in the baseset.
* The settings window is accessible from the game options, next to the baseset selection.
image
* The settings are stored in openttd.cfg in a separate section. When saving openttd.cfg the section is reset, so all items in the section are always written consistently. Future versions of OpenTTD can savely add more items to the section. 
```
[graphicsset]
name = OpenGFX2 Classic
shortname = 6F676678
extra_version = 0
extra_params = 0 1 0 0
```

* When downloading a new version of the baseset, the "min compatible version" from action14 is checked. The parameters are kept, if the versions are reported compatible. Otherwise all settings are reset to default values.

## Limitations

This PR works with OpenGFX2_Classic-0.1.

The current version 0.21 appears to be broken: It has parameters, but they do nothing. @zephyris any idea why?

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/base_media_base.h [hunk 1]
```diff
@@ -239,19 +238,36 @@ enum BlitterType {
...
 /** All data of a graphics set. */
 struct GraphicsSet : BaseSet<GraphicsSet, MAX_GFT, true> {
+private:
+	mutable std::unique_ptr<GRFConfig> extra_cfg; ///< Parameters for extra GRF
+public:
 	PaletteType palette;       ///< Palette of this graphics set
 	BlitterType blitter;       ///< Blitter of this graphics set
 
 	bool FillSetDetails(struct IniFile *ini, const std::string &path, const std::string &full_filename);
+	GRFConfig *GetExtraConfig(bool load=true) const;
```
Comment (PeterN) [inline src/base_media_base.h]: Spaces... `bool load = true`

File: src/gfxinit.cpp [hunk 2]
```diff
@@ -364,6 +354,47 @@ bool GraphicsSet::FillSetDetails(IniFile *ini, const std::string &path, const st
...
+	const GRFConfig* cfg = this->GetExtraConfig();
+	/* This check is more strict than the one for NewGRF Settings.
+	 * There are no legacy basesets with parameters, but without Action14 */
+	return cfg && !cfg->param_info.empty();
+}
+
+void GraphicsSet::CopyCompatibleConfig(const GraphicsSet &src)
+{
+	const GRFConfig* src_cfg = src.GetExtraConfig(false);
+	if (!src_cfg || src_cfg->num_params == 0) return;
```
Comment (PeterN) [inline src/gfxinit.cpp]: `src_cfg != nullptr`

File: src/gfxinit.cpp [hunk 3]
```diff
@@ -364,6 +354,47 @@ bool GraphicsSet::FillSetDetails(IniFile *ini, const std::string &path, const st
...
+	GRFConfig* dest_cfg = this->GetExtraConfig();
+	if (!dest_cfg || !dest_cfg->IsCompatible(src_cfg->version)) return;
```
Comment (PeterN) [inline src/gfxinit.cpp]: dest_cfg != nullptr

File: src/settings.cpp [hunk 4]
```diff
@@ -981,6 +985,36 @@ static bool DecodeHexText(const char *pos, uint8_t *dest, size_t dest_size)
...
 }
 
+/**
+ * Load BaseGraphics set selection and configuration.
+ */
+static void GraphicsSetLoadConfig(IniFile &ini)
+{
+	if (IniGroup *group = ini.GetGroup("misc"); group != nullptr) {
+		/* Load old setting first. */
+		if (IniItem *item = group->GetItem("graphicsset"); item != nullptr && item->value) BaseGraphics::ini_data.name = *item->value;
```
Comment (PeterN) [inline src/settings.cpp]: Interesting syntax inside if that I've not seen before.

File: src/settings.cpp [hunk 5]
```diff
@@ -1159,6 +1193,25 @@ static void SaveVersionInConfig(IniFile &ini)
 	group->GetOrCreateItem("ini_version").SetValue(std::to_string(INIFILE_VERSION));
 }
 
+/**
+ * Save BaseGraphics set selection and configuration.
+ */
+static void GraphicsSetSaveConfig(IniFile &ini)
+{
+	const GraphicsSet *used_set = BaseGraphics::GetUsedSet();
+	if (used_set) {
```
Comment (PeterN) [inline src/settings.cpp]: `if (used_set != nullptr) {`

or better,

`if (used_set == nullptr) return;`

File: src/settings.cpp [hunk 6]
```diff
@@ -1159,6 +1193,25 @@ static void SaveVersionInConfig(IniFile &ini)
...
+		ini.RemoveGroup("graphicsset");
+		IniGroup *group = ini.GetGroup("graphicsset");
+		group->GetOrCreateItem("name").SetValue(used_set->name);
+		group->GetOrCreateItem("shortname").SetValue(fmt::format("{:08X}", BSWAP32(used_set->shortname)));
+		const GRFConfig *extra_cfg = used_set->GetExtraConfig(false);
+		if (extra_cfg && extra_cfg->num_params > 0) {
```
Comment (PeterN) [inline src/settings.cpp]: As above.

File: src/settings_gui.cpp [hunk 7]
```diff
@@ -601,6 +602,16 @@ struct GameOptionsWindow : Window {
 				break;
 			}
 
+			case WID_GO_BASE_GRF_PARAMETERS: {
+				auto *used_set = BaseGraphics::GetUsedSet();
+				if (!used_set || !used_set->IsConfigurable()) break;
```
Comment (PeterN) [inline src/settings_gui.cpp]: `!= nullptr`

**Review by PeterN — CHANGES_REQUESTED** [review]
Nice

File: src/base_media_base.h [hunk 8]
```diff
@@ -239,19 +238,36 @@ enum BlitterType {
...
 	static MD5File::ChecksumResult CheckMD5(const MD5File *file, Subdirectory subdir);
 };
 
 /** All data/functions related with replacing the base graphics. */
 class BaseGraphics : public BaseMedia<GraphicsSet> {
 public:
+	/** Values loaded from config file. */
+	struct Ini {
+		std::string name;
+		uint32_t shortname;                 //!< unique key for base set
```
Comment (LordAro) [inline src/base_media_base.h]:
```suggestion
		uint32_t shortname;                 ///< unique key for base set
```

etc

File: src/gfxinit.cpp [hunk 9]
```diff
@@ -364,6 +354,47 @@ bool GraphicsSet::FillSetDetails(IniFile *ini, const std::string &path, const st
...
+			default: break;
+		}
+		FillGRFDetails(this->extra_cfg.get(), false, BASESET_DIR);
+	}
+	return this->extra_cfg.get();
+}
+
+bool GraphicsSet::IsConfigurable() const
+{
+	const GRFConfig* cfg = this->GetExtraConfig();
```
Comment (LordAro) [inline src/gfxinit.cpp]:
```suggestion
	const GRFConfig *cfg = this->GetExtraConfig();
```

etc

File: src/settings.cpp [hunk 10]
```diff
@@ -981,6 +985,36 @@ static bool DecodeHexText(const char *pos, uint8_t *dest, size_t dest_size)
...
+	}
+	if (IniGroup *group = ini.GetGroup("graphicsset", false); group != nullptr) {
+		/* Load new settings. */
+		if (IniItem *item = group->GetItem("name"); item != nullptr && item->value) BaseGraphics::ini_data.name = *item->value;
+		if (IniItem *item = group->GetItem("shortname"); item != nullptr && item->value && item->value->size() == 8) {
+			BaseGraphics::ini_data.shortname = BSWAP32(std::strtoul(item->value->c_str(), nullptr, 16));
+		}
+		if (IniItem *item = group->GetItem("extra_version"); item != nullptr && item->value) BaseGraphics::ini_data.extra_version = std::strtoul(item->value->c_str(), nullptr, 10);
+		if (IniItem *item = group->GetItem("extra_params"); item != nullptr && item->value) {
+			auto& extra_params = BaseGraphics::ini_data.extra_params;
```
Comment (LordAro) [inline src/settings.cpp]:
```suggestion
			auto &extra_params = BaseGraphics::ini_data.extra_params;
```

etc

File: src/settings.cpp (hunk 4 above)
Comment (frosch123) [reply src/settings.cpp]: https://youtu.be/tsG95Y-C14k?t=114

File: src/settings.cpp [hunk 11]
```diff
@@ -53,6 +54,9 @@
 ClientSettings _settings_client;
 GameSettings _settings_game;     ///< Game settings of a running game or the scenario editor.
 GameSettings _settings_newgame;  ///< Game settings for new games (updated from the intro screen).
+BaseGraphics::Ini BaseGraphics::ini_data;
+std::string BaseSounds::ini_set;
+std::string BaseMusic::ini_set;
```
Comment (PeterN) [inline src/settings.cpp]: Does it work to have these as static inline inside the class definitions (like how window hotkeys are set up), instead of dumped here?

File: src/settings.cpp (hunk 4 above)
Comment (LordAro) [reply src/settings.cpp]: https://en.cppreference.com/w/cpp/language/if#If_statements_with_initializer

**Comment by 2TallTyler** [comment]
The first two commits could be their own PR (or two). The second commit in particular would fix #8846 and #10549, if I'm reading everything correctly.

**Comment by zephyris** [comment]
Very nice, thank you!

Re. OpenGFX2 v0.1 vs v0.21 - it's my fault, see: https://github.com/zephyris/opengfx2/blob/7be1c3e7cf42d0eb1cebef52b162eb1763c99ba3/baseset/nml/extra/extra-header.pnml#L257-L297
I was trying to work out how to have the base set parameters (client side) and a settings grf (newgrf/server defined) coexist nicely. Without a way for me to set base set parameters for testing I gave up and hardcoded some defaults for now!

**Comment by PeterN** [comment]
Speaking of networking, presumably this doesn't introduce a method of baseset-desyncs?

**Comment by frosch123** [comment]
@PeterN There is already a loop hole for desyncs.
* If NewGRF check for presence of static NewGRF, the static NewGRF are disabled. (desync safe)
* If NewGRF check for presence of the baseset extra GRF, OpenTTD happily does that, and this can potentially desync.
* Some NewGRF do this to match sprites for different basesets.
* The same applies to reading parameters from other NewGRF: OpenTTD catches it for static GRF, but allows it for the baseset extra GRF.

For me it's a case of: Tolerate as long as noone causes desyncs. Fix on first occassion someone does something bad.

File: src/gfxinit.cpp [hunk 12]
```diff
@@ -347,6 +337,17 @@ void GfxLoadSprites()
...
+GraphicsSet::GraphicsSet()
+	: BaseSet<GraphicsSet, MAX_GFT, true>{}, palette{}, blitter{}
+{
+	// only here due to incomplete types
+}
+
+GraphicsSet::~GraphicsSet()
+{
+	// only here due to incomplete types
+}
```
Comment (PeterN) [inline src/gfxinit.cpp:349]: Why are these necessary now? Also, prefer member initialization of `palette` and `blitter` instead of initialization list.

File: src/settings.cpp [hunk 13]
```diff
@@ -1159,6 +1190,24 @@ static void SaveVersionInConfig(IniFile &ini)
...
+	if (used_set == nullptr) return;
+	ini.RemoveGroup("graphicsset");
+	IniGroup *group = ini.GetGroup("graphicsset");
+	group->GetOrCreateItem("name").SetValue(used_set->name);
+	group->GetOrCreateItem("shortname").SetValue(fmt::format("{:08X}", BSWAP32(used_set->shortname)));
+	const GRFConfig *extra_cfg = used_set->GetExtraConfig();
+	if (extra_cfg != nullptr && extra_cfg->num_params > 0) {
+		group->GetOrCreateItem("extra_version").SetValue(fmt::format("{}", extra_cfg->version));
+		group->GetOrCreateItem("extra_params").SetValue(GRFBuildParamList(extra_cfg));
+	}
```
Comment (PeterN) [inline src/settings.cpp]: Could perhaps do with some extra newlines to break it into more readable sections...

File: src/gfxinit.cpp (hunk 12 above)
Comment (frosch123) [reply src/gfxinit.cpp:349]: The destructor of `std::unique_ptr` needs a complete type, so it can call the destructor of the held type.
I do not want to include `newgrf_config.h` in `base_media_base.h`, so I have to put `~GraphicsSet` into some `.cpp` file.
For some reason MSVC also needs the constructor in a `.cpp` file.

Comment (PeterN) [reply src/gfxinit.cpp:349]: Ah, so something like this might be helpful:
`/* Explicit constructor/destructor required due to incomplete type used in unique_ptr type. */`

**Comment by glx22** [comment]
Test build available at https://cdn.openttd.org/openttd-branches/pr11347/

**Comment by zephyris** [comment]
Thanks @glx22. I've tested it with OpenGFX2 v0.1 and my latest OpenGFX2 dev build. Looks like it works exactly as expected.
* It's not 100% intuitive how to apply the changes, ie. that the settings change on exit from the settings window, but I think it is fine as is.
* The settings window is wider than before - presumably wider by the parameters button width? Should probably be narrowed down.
* I haven't tested the `action14` support.
* Looks like these changes also fix an old bug. Previously OpenTTD ignored default values for parameter values set in nml `grf {}` block, this behaves correctly now.

**Comment by PeterN** [comment]
>     * Looks like these changes also fix an old bug. Previously OpenTTD ignored default values for parameter values set in nml `grf {}` block, this behaves correctly now.

What's the issue number for the old bug? It might have been closed by related change #11348?

**Comment by zephyris** [comment]
It hasn't got a issue number, this comes from me trying to get OpenGFX2 parameters working as expected over the last week or so. It's only now, testing with this PR, that it's clear that it was OpenTTD defaulting to 0 for parameter values set in nml grf {} block (and not an nml syntax error or something else).

**Comment by PeterN** [comment]
Oh, okay... OpenTTD doesn't support baseset parameters without this change, so I'm not sure why you'd consider it a separate bug that baseset parameters were not supported...

**Comment by zephyris** [comment]
My naivety then :)  Moot point anyway, this PR handles it all.

**Review by PeterN — DISMISSED** [review]
Let's not let this go stale

File: src/settings.cpp [hunk 14]
```diff
@@ -1153,6 +1188,27 @@ static void SaveVersionInConfig(IniFile &ini)
...
+/**
+ * Save BaseGraphics set selection and configuration.
+ */
+static void GraphicsSetSaveConfig(IniFile &ini)
+{
+	const GraphicsSet *used_set = BaseGraphics::GetUsedSet();
+	if (used_set == nullptr) return;
+
+	ini.RemoveGroup("graphicsset");
+	IniGroup *group = ini.GetGroup("graphicsset");
```
Comment (PeterN) [inline src/settings.cpp]: `GetGroup()` will no longer create a group.

```suggestion
	IniGroup &group = ini.GetOrCreateGroup("graphicsset");
	group.Clear();
```

**Review by PeterN — APPROVED** [review]
